File: notale-v2/experiments/workflow-skills-next/sample-workbench/review/page-chart/vocal-register-timeline/tests/verify.py
#!/usr/bin/env python3
import json
import threading
from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import urlparse

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parents[1]
PAGES = ROOT / "pages"
SHOTS = ROOT / "screenshots" / "final"
CHROME = "/data1/home/zhuyifan/.cache/ms-playwright/chromium-1234/chrome-linux64/chrome"
SONG_COUNTS = {2: 100, 3: 100, 4: 200, 6: 100, 8: 100, 11: 100, 15: 80, 16: 5}
TOP_STATES = {13, 14, 17}


class QuietHandler(SimpleHTTPRequestHandler):
    def log_message(self, *_):
        pass


def check(condition, message):
    if not condition:
        raise AssertionError(message)


def main():
    SHOTS.mkdir(parents=True, exist_ok=True)
    server = ThreadingHTTPServer(("127.0.0.1", 0), partial(QuietHandler, directory=PAGES))
    threading.Thread(target=server.serve_forever, daemon=True).start()
    origin = f"http://127.0.0.1:{server.server_port}"
    report = {"status": "FAIL", "origin": origin, "states": [], "checks": []}
    try:
        with sync_playwright() as pw:
            browser = pw.chromium.launch(headless=True, executable_path=CHROME, args=["--no-sandbox"])
            context = browser.new_context(viewport={"width": 1600, "height": 900})
            page = context.new_page()
            errors, requests = [], []
            page.on("console", lambda m: errors.append(m.text) if m.type == "error" else None)
            page.on("pageerror", lambda e: errors.append(str(e)))
            page.on("request", lambda r: requests.append(r.url))
            page.goto(origin, wait_until="networkidle")
            page.wait_for_function("window.__vocalSample && document.querySelectorAll('.dot').length === 62")
            page.wait_for_timeout(2600)

            for state in range(18):
                page.evaluate("n => window.__vocalSample.go(n)", state)
                page.wait_for_timeout(2600 if state == 0 else 950)
                actual = page.evaluate("window.__vocalSample.state")
                check(actual == state, f"state {state} did not settle")
                check(page.locator("#copy").inner_text().strip(), f"state {state} has no claim")
                check(page.locator("#line").get_attribute("d").startswith("M"), "line path missing")
                if state in SONG_COUNTS:
                    check(page.locator(".song").count() == SONG_COUNTS[state], f"state {state} song count")
                    check("off" in page.locator("#chart").get_attribute("class"), f"state {state} chart visible")
                else:
                    check(page.locator("#selected circle").count() == (1 if state in {0, 1, 5, 7, 9, 10, 12} else 0), f"state {state} annotation")
                    check(page.locator(".dot.on").count() == (62 if state in TOP_STATES else 0), f"state {state} top dots")
                if state == 14:
                    check(page.locator(".dot.ring").count() == 1, "2019 emphasis ring missing")
                overflow = page.evaluate("""() => ({x:document.documentElement.scrollWidth-document.documentElement.clientWidth,
                    y:document.documentElement.scrollHeight-document.documentElement.clientHeight})""")
                check(overflow == {"x": 0, "y": 0}, f"state {state} overflow: {overflow}")
                page.screenshot(path=SHOTS / f"state-{state:02}.png")
                report["states"].append({"state": state, "songs": page.locator(".song").count()})

            page.evaluate("window.__vocalSample.go(0)")
            page.keyboard.press("ArrowRight")
            check(page.evaluate("window.__vocalSample.state") == 1, "ArrowRight")
            page.keyboard.press("ArrowLeft")
            check(page.evaluate("window.__vocalSample.state") == 0, "ArrowLeft")
            page.evaluate("window.__vocalSample.go(9)")
            page.keyboard.press("r")
            check(page.evaluate("window.__vocalSample.state") == 0, "keyboard reset")
            page.locator("#next").click()
            page.locator("#prev").click()
            check(page.evaluate("window.__vocalSample.state") == 0, "pointer navigation")
            page.evaluate("window.__vocalSample.go(12)")
            page.locator("#reset").click()
            first = page.evaluate("() => [window.__vocalSample.state, document.querySelector('#line').getAttribute('d'), document.querySelector('#copy').textContent]")
            page.evaluate("window.__vocalSample.go(12)")
            page.locator("#reset").click()
            second = page.evaluate("() => [window.__vocalSample.state, document.querySelector('#line').getAttribute('d'), document.querySelector('#copy').textContent]")
            check(first == second and first[0] == 0, "reset drift")
            report["checks"].append("18 states, keyboard, pointer navigation, and double reset")

            page.evaluate("window.__vocalSample.go(13)")
            page.set_viewport_size({"width": 1280, "height": 720})
            page.wait_for_timeout(100)
            check(page.evaluate("getComputedStyle(document.documentElement).getPropertyValue('--scale').trim()") == "0.8", "1280 scale")
            check(page.evaluate("document.documentElement.scrollWidth === document.documentElement.clientWidth"), "1280 overflow")
            page.screenshot(path=SHOTS / "viewport-1280x720.png")
            page.set_viewport_size({"width": 1600, "height": 900})
            check(page.evaluate("window.__vocalSample.state") == 13, "resize changed state")
            report["checks"].append("1600x900 and 1280x720 resize")

            external = [u for u in requests if urlparse(u).scheme in {"http", "https"} and not u.startswith(origin)]
            check(not external, f"external requests: {external}")
            check(not errors, f"browser errors: {errors}")
            report["requests"] = requests
            report["checks"].append("no console errors or external requests")

            reduced = browser.new_context(viewport={"width": 1600, "height": 900}, reduced_motion="reduce")
            rp = reduced.new_page()
            reduced_errors = []
            rp.on("console", lambda m: reduced_errors.append(m.text) if m.type == "error" else None)
            rp.on("pageerror", lambda e: reduced_errors.append(str(e)))
            rp.goto(origin, wait_until="networkidle")
            rp.wait_for_function("window.__vocalSample && document.querySelectorAll('.dot').length === 62")
            rp.evaluate("window.__vocalSample.go(13)")
            rp.wait_for_timeout(80)
            check(rp.locator(".dot.on").count() == 62, "reduced-motion final evidence")
            check(rp.evaluate("getComputedStyle(document.querySelector('#chart')).transitionDuration") == "0s", "reduced-motion CSS")
            check(not reduced_errors, f"reduced-motion errors: {reduced_errors}")
            rp.screenshot(path=SHOTS / "reduced-motion.png")
            reduced.close()
            report["checks"].append("reduced-motion final state")

            page.evaluate("window.__vocalSample.go(5); window.__vocalSample.teardown()")
            page.keyboard.press("ArrowRight")
            page.locator("#next").click()
            check(page.evaluate("window.__vocalSample.state") == 5, "teardown left listeners")
            report["checks"].append("teardown removes owned listeners and animation")
            context.close()
            browser.close()
        report["status"] = "PASS"
    except Exception as error:
        report["error"] = str(error)
        raise
    finally:
        server.shutdown()
        (ROOT / "tests" / "results.json").write_text(json.dumps(report, indent=2), encoding="utf-8")
    print(json.dumps(report, indent=2))


if __name__ == "__main__":
    main()

File: experiments/notale-v2/workflow-skills-next/sample-workbench/mini/page-chart/lenna-scroll-bars/tests/verify.mjs
import assert from "node:assert/strict";
import crypto from "node:crypto";
import fs from "node:fs";
import http from "node:http";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "/tmp/notale-playwright/node_modules/playwright/index.mjs";

const testDir = path.dirname(fileURLToPath(import.meta.url));
const sampleDir = path.dirname(testDir);
const pagesDir = path.join(sampleDir, "candidate/pages");
const types = { ".html":"text/html", ".js":"text/javascript", ".css":"text/css", ".csv":"text/csv" };
const server = http.createServer((request, response) => {
  const relative = request.url === "/" ? "index.html" : request.url.slice(1);
  const target = path.join(pagesDir, relative);
  response.setHeader("Content-Type", `${types[path.extname(target)]}; charset=utf-8`);
  response.end(fs.readFileSync(target));
});
await new Promise(resolve => server.listen(0, "127.0.0.1", resolve));
const url = `http://127.0.0.1:${server.address().port}/`;
const browser = await chromium.launch({ headless:true });
const errors = [];
const snapshot = page => page.evaluate(() => LennaMini.snapshot());

function aggregateCsv() {
  const text = fs.readFileSync(path.join(pagesDir, "assets/data/data.csv"), "utf8");
  const years = new Map();
  let records = 0;
  for (const line of text.trim().split(/\r?\n/).slice(1)) {
    const comma = line.lastIndexOf(",");
    const domain = line.slice(0, comma);
    const year = Number.parseInt(line.slice(comma + 1), 10);
    if (!Number.isInteger(year) || year < 1972) continue;
    const key = domain.endsWith(".org") ? ".org" : domain.endsWith(".edu") ? ".edu" :
      domain.endsWith(".com") ? ".com" : "other";
    const row = years.get(year) || { year, value:0, ".org":0, ".edu":0, ".com":0, other:0 };
    row.value += 1;
    row[key] += 1;
    years.set(year, row);
    records += 1;
  }
  return { records, rows:[...years.values()].sort((a, b) => a.year - b.year) };
}

async function open(options = {}) {
  const context = await browser.newContext({
    viewport:options.viewport || { width:1600, height:900 },
    reducedMotion:options.reduced ? "reduce" : "no-preference"
  });
  if (options.noData) await context.route("**/assets/data/data.csv", route => route.abort());
  const page = await context.newPage();
  page.on("pageerror", error => errors.push(error.message));
  await page.goto(url, { waitUntil:"networkidle" });
  if (options.noData) await page.waitForSelector("#fallback:not([hidden])");
  else await page.waitForFunction(() => LennaMini.snapshot().ready);
  return { context, page };
}

try {
  const canonical = aggregateCsv();
  assert.equal(canonical.records, 4801);
  assert.equal(canonical.rows.length, 45);
  assert.deepEqual(canonical.rows[0], { year:1972, value:2, ".org":1, ".edu":0, ".com":1, other:0 });
  assert.deepEqual(canonical.rows.find(row => row.year === 1995),
    { year:1995, value:287, ".org":222, ".edu":14, ".com":34, other:17 });
  assert.deepEqual(canonical.rows.at(-1),
    { year:2021, value:252, ".org":36, ".edu":0, ".com":148, other:68 });

  for (const viewport of [{ width:1600, height:900 }, { width:1280, height:720 }]) {
    const run = await open({ viewport });
    assert.deepEqual(await snapshot(run.page),
      { ready:true, state:"1972", stateIndex:0, destroyed:false });
    const layout = await run.page.evaluate(() => ({
      width:innerWidth, height:innerHeight,
      scrollWidth:document.documentElement.scrollWidth,
      scrollHeight:document.documentElement.scrollHeight,
      stage:[stage.getBoundingClientRect().width, stage.getBoundingClientRect().height],
      bars:document.querySelectorAll(".bar").length,
      segments:document.querySelectorAll(".segment").length,
      visible:[...document.querySelectorAll(".bar")].filter(node => node.style.display !== "none").length
    }));
    assert.deepEqual([layout.scrollWidth, layout.scrollHeight], [layout.width, layout.height]);
    assert.deepEqual(layout.stage, [layout.width, layout.height]);
    assert.deepEqual([layout.bars, layout.segments, layout.visible], [45, 143, 1]);
    await run.page.screenshot({ path:path.join(testDir, `${viewport.width}-1972.png`) });
    await run.context.close();
  }

  const motion = await open();
  await motion.page.evaluate(() => { window.peakBarIdentity = document.querySelector(".bar[data-year='1995']"); });
  await motion.page.focus("#next");
  await motion.page.keyboard.press("Enter");
  await motion.page.waitForTimeout(420);
  const intermediate = await motion.page.evaluate(() => {
    const visible = [...document.querySelectorAll(".bar")].filter(node => node.style.display !== "none");
    return {
      state:LennaMini.snapshot().state,
      opacity:visible.map(node => Number(getComputedStyle(node).opacity)),
      delays:visible.map(node => Number(node.dataset.revealDelay || 0)),
      sameNode:peakBarIdentity === document.querySelector(".bar[data-year='1995']")
    };
  });
  assert.equal(intermediate.state, "1995");
  assert.equal(intermediate.sameNode, true);
  assert.ok(intermediate.opacity.some(value => value > 0 && value < 1));
  assert.ok(intermediate.opacity.some(value => value === 0));
  assert.deepEqual(intermediate.delays.slice(1),
    Array.from({ length:18 }, (_, index) => index * 80));
  await motion.page.waitForTimeout(1500);
  assert.equal(await motion.page.locator(".bar.current").getAttribute("data-year"), "1995");
  assert.match(await motion.page.locator("#copy").innerText(), /1995.*287/);
  assert.equal(await motion.page.locator(".x-axis text.current").textContent(), "1995");
  assert.equal(await motion.page.locator(".x-axis text").filter({ hasText:"1996" }).isVisible(), false);
  await motion.page.screenshot({ path:path.join(testDir, "1600-1995.png") });

  await motion.page.click("#next");
  await motion.page.waitForTimeout(500);
  assert.equal((await snapshot(motion.page)).state, "domains");
  assert.equal(await motion.page.locator(".segment").count(), 143);
  assert.equal(await motion.page.locator(".segment").first().isVisible(), true);
  assert.equal(await motion.page.locator(".domain-key").isVisible(), true);
  const peakGeometry = await motion.page.evaluate(() => {
    const segments = [...document.querySelectorAll(".segment[data-year='1995']")];
    const bar = document.querySelector(".bar[data-year='1995']");
    return {
      categories:segments.map(node => node.dataset.domain),
      stackHeight:segments.reduce((sum, node) => sum + Number(node.getAttribute("height")), 0),
      barHeight:Number(bar.getAttribute("height"))
    };
  });
  assert.deepEqual(peakGeometry.categories, [".org", ".edu", ".com", "other"]);
  assert.ok(Math.abs(peakGeometry.stackHeight - peakGeometry.barHeight) < 1e-8);
  await motion.page.screenshot({ path:path.join(testDir, "1600-domains.png") });

  await motion.page.click("#previous");
  assert.equal((await snapshot(motion.page)).state, "1995");
  await motion.page.click("#reset");
  await motion.page.click("#reset");
  assert.deepEqual(await snapshot(motion.page),
    { ready:true, state:"1972", stateIndex:0, destroyed:false });
  assert.equal(await motion.page.locator(".bar").filter({ visible:true }).count(), 1);
  await motion.context.close();

  const reduced = await open({ reduced:true, viewport:{ width:1280, height:720 } });
  await reduced.page.click("#next");
  const reducedOpacity = await reduced.page.locator(".bar").evaluateAll(nodes => nodes
    .filter(node => node.style.display !== "none").map(node => Number(getComputedStyle(node).opacity)));
  assert.ok(reducedOpacity.every(value => value === 1));
  await reduced.page.click("#next");
  const segmentOpacity = await reduced.page.locator(".segment").evaluateAll(nodes =>
    nodes.map(node => Number(getComputedStyle(node).opacity)));
  assert.ok(segmentOpacity.every(value => value === 1));
  await reduced.page.screenshot({ path:path.join(testDir, "1280-reduced-domains.png") });
  await reduced.context.close();

  const fallback = await open({ noData:true });
  assert.equal(await fallback.page.locator("#fallback").isVisible(), true);
  assert.equal(await fallback.page.locator(".bar").count(), 0);
  assert.match(await fallback.page.locator("#fallback").innerText(), /data\.csv.*不会用占位数值/s);
  await fallback.context.close();

  const lifecycle = await open();
  await lifecycle.page.click("#next");
  await lifecycle.page.evaluate(() => LennaMini.destroy());
  const disposed = await snapshot(lifecycle.page);
  assert.equal(disposed.destroyed, true);
  await lifecycle.page.click("#next");
  await lifecycle.page.click("#reset");
  assert.deepEqual(await snapshot(lifecycle.page), disposed);
  assert.equal(await lifecycle.page.evaluate(() =>
    [...document.querySelectorAll(".bar,.segment")].some(node => Boolean(node.__transition))), false);
  await lifecycle.page.setViewportSize({ width:1280, height:720 });
  await lifecycle.page.reload({ waitUntil:"networkidle" });
  await lifecycle.page.waitForFunction(() => LennaMini.snapshot().ready);
  assert.deepEqual(await snapshot(lifecycle.page),
    { ready:true, state:"1972", stateIndex:0, destroyed:false });
  await lifecycle.context.close();

  const dataHash = crypto.createHash("sha256")
    .update(fs.readFileSync(path.join(pagesDir, "assets/data/data.csv"))).digest("hex");
  assert.equal(dataHash, "18dbc5fd90fbb6c536f5d2129ac0d093d574708e98dbc274a516169cf97ab249");
  assert.deepEqual(errors, []);
  console.log(JSON.stringify({ passed:true, checks:[
    "4,801 records aggregate to 45 annual rows and 143 stable stack segments",
    "1972, 1995 and domains states retain exact values and shared geometry",
    "80ms reveal has observable intermediate frames and stable bar identity",
    "two viewports, reduced motion, double reset and truthful data fallback",
    "transition interruption, Abort teardown, resize and clean reload"
  ] }, null, 2));
} finally {
  await browser.close();
  server.close();
}
